core/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java

```diff
@@ -2914,6 +2928,108 @@ public void testSequenceIDs() throws Exception {
         searchResult.close();
     }
 
+    public void testSequenceNumberAdvancesToMaxSeqNoOnEngineOpen() throws IOException {
+        engine.close();
+        final long maxSeqNo = randomIntBetween(16, 32);
+        final long localCheckpoint =
+            rarely() ? SequenceNumbersService.NO_OPS_PERFORMED : randomIntBetween(0, Math.toIntExact(maxSeqNo));
+        final long globalCheckpoint =
+            localCheckpoint == SequenceNumbersService.NO_OPS_PERFORMED || rarely()
+                ? SequenceNumbersService.UNASSIGNED_SEQ_NO
+                : Math.toIntExact(localCheckpoint);
+        Engine initialEngine = null;
+        try {
+            initialEngine = createEngine(
+                defaultSettings,
+                store,
+                primaryTranslogDir,
+                newMergePolicy(),
+                null,
+                () -> new SequenceNumbersService(shardId, defaultSettings, maxSeqNo, localCheckpoint, globalCheckpoint) {
+                    @Override
+                    public long generateSeqNo() {
+                        if (rarely()) {
+                            // force skipping a sequence number
+                            super.generateSeqNo();
+                        }
+                        return super.generateSeqNo();
+                    }
+                });
+            initialEngine.flush(true, true);
+            final int docs = randomIntBetween(1, 32);
+            for (int i = 0; i < docs; i++) {
+                final String id = Integer.toString(i);
+                final ParsedDocument doc = testParsedDocument(id, id, "test", null, testDocumentWithTextField(), SOURCE, null);
+                initialEngine.index(new Engine.Index(newUid(id), doc));
+            }
+        } finally {
+            IOUtils.close(initialEngine);
+        }
+
+        try (final Engine recoveringEngine =
+                 new InternalEngine(copy(initialEngine.config(), EngineConfig.OpenMode.OPEN_INDEX_AND_TRANSLOG))) {
+            // there might be gaps in the translog; the best that we can say is that the local checkpoint is at least as far the max
+            // sequence number
+            // TODO: tighten this assertion after gaps in the translog are addressed
+            assertThat(recoveringEngine.seqNoService().getLocalCheckpoint(), greaterThanOrEqualTo(maxSeqNo));
+        }
+    }
+
+    public void testOutOfOrderSequenceNumbersWithVersionConflict() {
```

Comment:
++. Maybe also add a sanity check that a get on the doc at the end gives us what we expect? (deleted or found)